JSON equality operator

LAJW opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My PR is restricted to a single feature or bugfix.


<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
JSON equality operator and tests

Because it'd be useful to be able to compare whole JSON objects (for
instance in unit tests)

thomasspriggs commented on src/util/json.cpp:
```diff
@@ -162,3 +164,49 @@ void jsont::swap(jsont &other)
+    const auto &right_object = static_cast<const json_objectt &>(right);
+    const auto yes = [](const std::pair<std::string, jsont> &) { return true; };
+    const size_t left_size = narrow<size_t>(
+      std::count_if(left_object.begin(), left_object.end(), yes));
```
:grey_question: Can this be done using `std::count`, so that you don't need to define `yes`?

LAJW replied:
`std::count` counts occurrences of a specified element, so no.

thomasspriggs reviewed (approved):
Looks useful for writing unit tests.

smowton commented on src/util/json.cpp:
```diff
@@ -162,3 +164,49 @@ void jsont::swap(jsont &other)
+    const auto &right_object = static_cast<const json_objectt &>(right);
+    const auto yes = [](const std::pair<std::string, jsont> &) { return true; };
+    const size_t left_size = narrow<size_t>(
+      std::count_if(left_object.begin(), left_object.end(), yes));
```
`std::distance(left_object.begin(), left_object.end())`

smowton replied:
Or since `object` is `std::map`, just use `.size()`. At present since it is `map` you could also use `zip` and safely assume the keys occur in the same order, if guarded by a static-assert checking that `typeof(object)` really is an (ordered) map.

LAJW replied:
`json_objectt` is not a `map`, it just has some `map`-like methods exposed. `size` sadly is not one of them.
`zip` lives in test-gen AFAIK.
I don't want to rely on the order of elements in the map. We've been shifting towards using `unordered_map` in many places, and JSON looks like it could be one of them.

smowton replied:
`json_objectt` is not, but its member `object` is. `zip` lives in https://github.com/diffblue/cbmc/blob/develop/src/util/range.h

smowton replied:
Sure, but if you guard with an appropriate `static_assert` (ideally there would be some type trait `is_ordered`) then you could be sure this code will be changed as/when we do that

LAJW replied:
`object` is a protected property.

There isn't an `is_ordered` trait. Being "ordered" is a runtime property.

If I didn't have to worry about future proofing - `zip` is unnecessary - `std::equal(b1, e1, b2)` is the tool for this job.

smowton replied:
At least expose a `size` property on `json_objectt` rather than hacking around it with the existing interface. The ordered vs. unordered thing I'll pass on.

Commit:
Remove linter braces rule

Braces are already checked by clang-format, and this rule doesn't
allow me to use initializer list constructors/uniform initialization
in unit tests

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: cfe09dccb93f4a913a02c5bc7baabc277c58cf6e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119292309
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on src/util/json.cpp:
```diff
@@ -7,8 +7,10 @@ Author: Daniel Kroening, kroening@kroening.com
 \*******************************************************************/
 
 #include "json.h"
+#include <util/narrow.h>
```
For consistency: `#include "narrow.h"`

tautschnig commented on src/util/json.cpp:
```diff
@@ -162,3 +164,48 @@ void jsont::swap(jsont &other)
+  case jsont::kindt::J_NUMBER:
+    return left.value == right.value;
+  case jsont::kindt::J_STRING:
+    return left.value == right.value;
```
Overall design question: shouldn't we do this more in a visitor-style where each call deriving from `jsont` defines each own `operator==` and the switch-case here would only do the downcasts and then invoke the more specific equality operators?

LAJW replied:
Yes, except `jsont` much like `irept` doesn't allow any virtual functions. Therefore the visitor pattern is not applicable.

tautschnig replied:
Of course, hence me saying "visitor-style." But also my comment was intended as a thought, not any kind of hard requirement.

tautschnig commented on src/util/json.cpp:
```diff
@@ -162,3 +164,48 @@ void jsont::swap(jsont &other)
+    const size_t left_size =
+      narrow<size_t>(std::distance(left_object.begin(), left_object.end()));
+    const size_t right_size =
+      narrow<size_t>(std::distance(right_object.begin(), right_object.end()));
```
Why do those conversions when afterwards you're only going to use the values for comparing them to each other? I'd go with `const auto left_size = std::distance ...`

tautschnig commented on src/util/json.cpp:
```diff
@@ -162,3 +164,48 @@ void jsont::swap(jsont &other)
+      left_object.begin(),
+      left_object.end(),
+      [&](const std::pair<std::string, jsont> &pair) {
+        return right_object[pair.first] == pair.second;
```
That's an unnecessary lookup when instead you could just use two iterators over the two containers.

LAJW replied:
Yeah, but if someone decides to swap `std::map` for `std::unordered_map`, it could be difficult to debug, so this is more stable/doesn't rely on any invariants.

tautschnig replied:
I do wonder which containers would exhibit differences in strict equality testing? But I suppose it is conceivable that two `std::unordered_map` had their elements constructed in a different order, yet should compare equal. So, yes, keep it as is.

smowton replied:
Personally I still favour 
```
// This relies on ordered maps:
static_assert(std::is_same<typeof(jsont::object), std::map<std::string, jsont>>::value)
```

tautschnig commented on unit/json/json_parser.cpp:
```diff
@@ -133,3 +133,110 @@ SCENARIO("Loading JSON files")
   }
 }
+
+TEST_CASE("json equality", "[core][test-gen-util][state][type]")
```
I actually don't know what those [...][...] parts are good for at all, but this particular entry ("test-gen-util") strikes me as information of no value in this repository.

tautschnig commented on scripts/cpplint.py:
```diff
@@ -3974,13 +3974,6 @@ def CheckBracesSpacing(filename, clean_lines, linenum, nesting_state, error):
-    if (not Match(r'^[\s}]*[{.;,)<>\]:]', trailing_text)
-        and not _IsType(clean_lines, nesting_state, leading_text)):
-      error(filename, linenum, 'whitespace/braces', 5,
-            'Missing space before {')
```
Maybe removing this is the right thing to do, but then we should remove all the code computing `leading_text`, `trailing_text` above. Or maybe you should just use `NOLINTNEXTLINE(whitespace/braces)` in the test.

LAJW replied:
`clang-format` already checks for positions and spacing of curly braces and parentheses. The problem stems from the fact this isn't aware of initializer lists/uniform initialization.
```c++
std::vector<std::pair<int, int>>{{1, 2}, {3, 4}, {5, 6}};
```
In the days of C++98, the above wasn't a valid syntax. Curly braces would appear only after `if` /`for` statements, as block scopes or surround function bodies. The linter expects this to still be the case.

tautschnig replied:
Ok, in which case the linter should be cleaned properly so as not to leave in place unused code.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1a7ca85632241d0c7a41d6c4afd412b5c36dd05a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119297176

Commit:
Add size property to json_objectt

LAJW commented:
Done!
 - Exposed `jsont::object::size` as `json_objectt::size`
 - Used that instead of `std::distance`
 - Removed dead linter code

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2f63cb81329a35b4b75a7de7eec980adff4f8bfd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119350287
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

thomasspriggs commented on src/util/json.cpp:
```diff
@@ -162,3 +163,44 @@ void jsont::swap(jsont &other)
+    return left.value == right.value;
+  case jsont::kindt::J_ARRAY:
+  {
+    const auto &left_array = static_cast<const json_arrayt &>(left);
```
:information_source: There is a function called `to_json_array` at the bottom of this file, which you could re-use instead of adding your own static casts.

LAJW replied:
`static_cast` is a feature of C++, and not "my" cast. `to_json_array` checks for kind, which in this case is already checked. We'd be doing extra work, so no.

thk123 reviewed (approved):
Code owner review of the cpplint change only (though the other changes lgtm)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c70cbabf2ed0daeb86d385306d6fb38f9a0bb8eb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119476532
